qa/rpc-tests/test_framework/cashlib/cashlib.py

```diff
@@ -160,6 +162,178 @@ def spendscript(*data):
             assert 0, "cannot push %d bytes" % l
     return b"".join(ret)
 
+class ScriptMachine:
+    STACK = 0
+    ALTSTACK = 1
+
+    SCRIPT_VERIFY_P2SH = 1
```

Comment:
wonder if there's a way to poke into `src/script/interpreter.h` via some kind of C++ magic binding rather than duplicate all the values here. if memory serves @awemany had a PR to let python access C++ defined data structure but I could be wrong.

Reply:
@sickpig: Yes, SWIG does that for you. @gandrewstone, have you looked into using SWIG instead?

I have PR #964 open (which is admittedly in bad shape at the moment), but that is a nice demo of what can all be automated using SWIG.

Reply:
Let us examine this PR (making a Script isolated virtual machine) separately from the idea of moving libbitcoincash to SWIG.  In the Python layer, the underlying is wrapped in a class so can be swapped out.  As this library grows, its more and more looking like we'll need to move to SWIG.  Perhaps @awemany could show us what he's done in #964 in a dev meeting.

I've used SWIG extensively in the past.  It has its obvious pluses, minuses are a bit more subtle, so let me first note that the direct .so techniques and boost::python were created AFTER SWIG which was first released in 1995.  Issuers are: byzantine swig-files, controlling what APIs you don't want, underlying API changes force swig-file changes and few devs understand swig-files, memory management, and what you end up with in the higher language you generally want to wrap anyway. This means rather than "SCRIPT_VERIFY_P2SH=1" you end up with "SCRIPT_VERIFY_P2SH = ccode.SCRIPT_VERIFY_P2SH".  This doesn't change LOC, but reduces maintenance and bugs if definitions could change.  But given the stability of certain aspects of Bitcoin Cash, these defines are unlikely to change...

Essentially, SWIG doesn't really deliver on its promise to provide an API to higher level languages without touching the underlying, or its promise to provide a beautiful, directly usable API to higher level languages.  But what it DOES give you is many fewer changes to the underlying and allows you to create a much simpler higher-level language wrapper.

Reply:
>  This means rather than "SCRIPT_VERIFY_P2SH=1" you end up with "SCRIPT_VERIFY_P2SH = ccode.SCRIPT_VERIFY_P2SH". This doesn't change LOC, but reduces maintenance and bugs if definitions could change. But given the stability of certain aspects of Bitcoin Cash, these defines are unlikely to change...